### ISSUE TITLE
Failed Decryption with alternative default application selected

### ISSUE THREAD

**Issue by D0L0S** [SMSSecure/SMSSecure#47]
Android Version: Lollipop (5.0.1)
Default SMS Application: Hangouts

Details: 
When key exchange is attempted from within the application with another application set as the default sms application the key exchange does not complete successfully.

Steps:
1. Set Default SMS application to application other than SMSSecure
    - Settings -> More -> Default SMS app
    - Select Hangouts
2. Within message thread select "Start Secure Session"
    - Keys are exchanged however messages receieved are not decrypted.

**Comment by BastienLQ** [comment]
It works on my phone with AOSP SMS app as the default SMS application. I don't have Hangouts, but that might be a Hangouts bug.

**Comment by D0L0S** [comment]
Pertentially, I have a Nexus5, I will install some other SMS applocations are report back

**Comment by BastienLQ** [comment]
In `com.google.android.talk` (Hangouts):
```
        <receiver android:name="com.google.android.apps.hangouts.sms.SmsReceiver" android:permission="android.permission.BROADCAST_SMS" android:enabled="false">
            <intent-filter android:priority="2147483647">
                <action android:name="android.provider.Telephony.SMS_RECEIVED" />
            </intent-filter>
        </receiver>
```

In `com.android.mms` (AOSP SMS app):
```
        <receiver android:name=".transaction.PrivilegedSmsReceiver"
            android:permission="android.permission.BROADCAST_SMS">
            <intent-filter>
                <action android:name="android.provider.Telephony.SMS_DELIVER" />
            </intent-filter>
        </receiver>
```

In SMSSecure: https://github.com/SMSSecure/SMSSecure/blob/master/AndroidManifest.xml#L237-247

So it may be related to Hangouts. @pR0Ps, can you confirm this?

**Comment by D0L0S** [comment]
I can confirm I am only seeing this when Hangouts is the default sms application. It appears that both the secure session setup and teardown messages SMSSecure is sending are not being passed from Hangouts to the app. 
I have confirmed this between two Nexus 5 devices both running 5.0.1

**Comment by pR0Ps** [comment]
I tried it on my device (KitKat) and didn't ever see the encrypted messages or the key exchange messages get missed by SMSSecure. However, I did see the session teardown messages get ignored (which is a definite bug that I'll look into).


The way listening for SMS's is done on KitKat+ is that the default SMS application gets the `SMS_DELIVERED_ACTION` intent, and every app (including the default one), gets a `SMS_RECEIVED_ACTION` intent.

The priority value only does something on Jelly Bean and earlier (as far as I know) and controls who gets the `SMS_DELIVERED_ACTION` first. That app can then choose to stop it propagating or let the next app in line see it (and so on).

The way SMSSecure handles this is to listen for all `SMS_RECEIVED_ACTION`s and only do something with them if it's not the default SMS app on KitKat+ and if it detects that the message is a key exchange or an encrypted message. The `SMS_DELIVERED_ACTION` will take care of the message if it's set as the default SMS app.

This logic can be seen [here](https://github.com/SMSSecure/SMSSecure/blob/master/src/org/smssecure/smssecure/service/SmsListener.java).

Unless something in the encrypted message detection is broken or Lollipop drastically changed the was SMS is handled (again), key exchanges and encrypted messages should always work in SMSSecure, regardless of if it's the default SMS app of not.

Unfortunately I don't have a 5.0.1 device to test this on at the moment. I'll see if I can borrow one and try it out.
